### PR TITLE
Add Skip Processing Flag for DSR Execution

### PR DESCRIPTION
**Pull request by pattisdr**
Closes https://github.com/ethyca/fidesplus/issues/1059
https://github.com/ethyca/fides/issues/2311

### Description Of Changes

Allow collections and endpoints (for saas connections) to be skipped via a `skip_processing` flag during DSR execution.  This is for **DSR processing only**, not consent propagation to third party systems.

- Add a `skip_processing` flag for collections > fides_meta in dataset yaml files. If present, that collection will be omitted from the execution graph when processing a DSR.  

```yaml
dataset:
  - fides_key: postgres_example_test_dataset
    name: Postgres Example Test Dataset
    description: Example of a Postgres dataset containing a variety of related tables like customers, products, addresses, etc.
    collections:
      - name: address
        fides_meta:
          skip_processing: True
```

- For completeness, a similar flag can be added to the saas config at the endpoint level if needed:
```yaml
saas_config:
  fides_key: saas_connector_example
  name: SaaS Example Config
  type: custom
  description: A sample schema representing a SaaS for Fides
  version: 0.0.1

  endpoints:
    - name: skipped_collection
      skip_processing: True
      requests:
        read:
          method: GET
          path: /v1/misc_endpoint/<list_id>
          param_values:
            - name: list_id
              references:
                - dataset: saas_connector_example
                  field: users.list_ids
                  direction: from
```

#### For allowing collections to be skipped for databases:
* [ ] Fideslang requirement update, adds "skip_processing" Optional boolean under CollectionMeta
* [ ] Updates `convert_dataset_to_graph` which is called with `Dataset.get_graph` to omit collections with `skip_processing` set to True under collection fides meta.  This would be relevant for a yaml file associated with a database.

#### For allowing collections (endpoints) to be skipped in saas connectors:
* [ ] Updates `merge_datasets`, also called by `Dataset.get_graph` which is relevant for Saas configs only.  Combining the saas config and the dataset config into a single graph.  This also removes collections with `skip_processing`. 
* [ ] Modifies `SaaSConfig.get_graph` to pass on the value of `skip_processing` when building the Collection so it can later be used to remove the collection in `merge_datasets`. We don't want to remove the collection at this point, but do this later in `merge_datasets`, as the existence of a skipped collection in the dataset yaml would cause it to be added back in the merge.

#### For improving ease of restart
* [ ] Fixes an earlier issue where we prevented a graph from being restarted if we didn't have a saved collection and step.  We originally only allowed graphs to be restarted if they failed within the graph.   If there's no saved checkpoint, restart from the beginning.  Added this here so if a DSR fails because customer added "skip_processing" to a collection that was needed, they can make adjustments and then resume from the beginning.

### Steps to Confirm

* [x] `nox -s dev -- shell postgres`
* [x] Separate window, `cd fides/clients, turbo run dev`
* [x] Go to "Add Systems" > Add a system
* [x] Add a System name, click Save
* [x] Go to integrations tab, select PostgreSQL
* [x] Enter in postgres credentials for our test environment (these are _not_ sensitive,  fides ships with these secrets, this is just for testing locally in the postgres database running in docker on your local machine):
    - host: host.docker.internal
    - port: 6432
    - username, p/w postgres, postgres
    - database: postgres_example
 * [x]  Click Save
 * [x] Dataset field will have appeared.Locate the file: `src/fides/data/sample_project/sample_resources/postgres_example_test_dataset.yml` within fides.  Modify the visit collection to have a skip_processing flag like so:
 ```yaml
      - name: visit
        fides_meta:
          skip_processing: true
        fields:
          - name: email
            data_categories: [user.contact.email]
            fides_meta:
              identity: email
              data_type: string
          - name: last_visit
            data_categories: [system.operations]
```
* [x] Click the little pen icon by dataset, and copy and past this dataset (removing "dataset" and bullet point beside `  - fides_key: postgres_example_test_dataset`
* [x] Click "Save"
* [x] Click Test Integration and verify that it's green
* [x] Go to the privacy center, and submit an access request for `customer-1@example.com`. Check your `fides_uploads` directory and confirm that the `visit` collection is not in the upload.  It is similarly absent from the event logs for that privacy request.
* [x] Delete the integration and then navigate to manage datasets and delete that too. ( I cannot figure out how to update yaml via the UI). Create another integration, and another dataset, but this time, add `fides_meta > skip_processing` under the `address` collection which is needed for DSR processing
* [x] Submit another access request. This errors very early because `address` does not exist.  Privacy Request should be in an error state.
* [x] Delete integration, delete dataset.  Add a new integration and a new dataset where address collection is no longer skipped
* [x] Go to the privacy center, reprocess the Errored one.  It should complete successfully. Previously, you couldn't reprocess this.

### Pre-Merge Checklist

* [ ] All CI Pipelines Succeeded
* Documentation:
  * [ ] documentation complete, [PR opened in fidesdocs](https://github.com/ethyca/fidesdocs/pulls)
  * [ ] documentation [issue created in fidesdocs](https://github.com/ethyca/fidesdocs/issues/new/choose)
* [ ] Issue Requirements are Met
* [ ] Relevant Follow-Up Issues Created
* [ ] Update `CHANGELOG.md`
* [ ] For API changes, the [Postman collection](https://github.com/ethyca/fides/blob/main/docs/fides/docs/development/postman/Fides.postman_collection.json) has been updated

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -12,7 +12,7 @@ expandvars==0.9.0
 fastapi[all]==0.89.1
 fastapi-caching[redis]==0.3.0
 fastapi-pagination[sqlalchemy]==0.11.4
-fideslang==1.4.5
+fideslang @ git+https://github.com/ethyca/fideslang.git@ee1323ccdeca69c9c80c1d3c1423486cca74ba19#egg=fideslang
```
Comment (pattisdr) [inline requirements.txt:15]: This is temporarily pinned here https://github.com/ethyca/fideslang/compare/fidesplus_1059_dsr_flag_1_4_5?expand=1, adding the new "skip_processing" flag to CollectionMeta, but based off of the 1.4.5 tag until the fideslang 2.0 work is merged

**Comment by pattisdr** [comment]
Ah some tests to address first!

File: requirements.txt (hunk 1 above)
Comment (pattisdr) [reply requirements.txt:15]: Argh okay we just released fideslang 1.4.6 today. I am just going to hold off on getting this PR updated with main for now and wait for the fideslang 2.0 dust to settle hopefully.

**Review by galvana — CHANGES_REQUESTED** [review]
Nice easy to follow change! I just left some comments around the docstrings and logger statements that might help with troubleshooting in the future

File: src/fides/api/api/v1/endpoints/privacy_request_endpoints.py [hunk 2]
```diff
@@ -1057,18 +1056,13 @@ def bulk_restart_privacy_request_from_failure(
...
-            failed.append(
-                {
-                    "message": f"Cannot restart privacy request from failure '{privacy_request.id}'; no failed step or collection.",
-                    "data": {"privacy_request_id": privacy_request_id},
-                }
-            )
-            continue
 
         succeeded.append(
             _process_privacy_request_restart(
```
[the remark this answers is not in the document]
Comment (pattisdr) [reply src/fides/api/api/v1/endpoints/privacy_request_endpoints.py:1061]: good callout!

File: tests/ops/util/test_dataset_yaml.py [hunk 3]
```diff
@@ -341,3 +342,81 @@ def test_return_all_elements_specified_on_non_array_field():
...
+        ds = Dataset.parse_obj(dataset)
+        converted_dataset = convert_dataset_to_graph(ds, "ignore")
+        assert len(converted_dataset.collections) == 1
+        assert converted_dataset.collections[0].name == "b_collection"
+        assert converted_dataset.collections[0].skip_processing is False
+
+        assert DatasetGraph(converted_dataset)
+
+    def test_invalid_collection_skip_processing(self):
+        """Skipping a collection that shouldn't be skipped is not picked in convert_dataset_to_graph, but
```
[the remark this answers is not in the document]
Comment (pattisdr) [reply tests/ops/util/test_dataset_yaml.py:407]: Yes! I don't think we have all the information to necessarily flag this in `convert_dataset_to_graph` as this is just looking at one dataset at a time.  A collection in Dataset A may be referencing a collection in Dataset B, which is not available within that function. But instantiating DatasetGraph has you pass in all the Datasets that are relevant for DSR processing.

I see I have a typo here which makes it read in an odd way, so I'll adjust that.

File: requirements.txt (hunk 1 above)
Comment (pattisdr) [reply requirements.txt:15]: To pull in main and retest I have a new commit based off of 1.4.6 here in the requirements: https://github.com/ethyca/fideslang/pull/163.  This is still temporary until fideslang 2.0 is merged, hopefully tomorrow!

**Comment by pattisdr** [comment]
I think this is ready for another pass @galvana. Won't be merging until after fideslang 2.0 migration changes are in and I get another fideslang migration released.

**Comment by pattisdr** [comment]
@galvana the prerequisite fideslang release is out now, ready for a final review, hoping to merge today!

**Review by galvana — APPROVED** [review]
Thanks for making the recommended changes!

**Comment by pattisdr** [comment]
yay thank you for your review @galvana 🎉 🎉 🎉

**Pull request merged**